### PR TITLE
Add support for NVTX in cuda traces using existing instrumentation

### PR DESCRIPTION
**Pull request by XL64**
Added support for CUDA NvToolsExt tracing reusing already present instrumentation.
image

### CODE REVIEW

**Review by sframba — APPROVED** [review]

File: src/cmake/thirdparty/SetupGeosxThirdParty.cmake [hunk 1]
```diff
@@ -644,4 +644,12 @@ if(NOT ENABLE_${upper_LAI})
 endif()
 option(GEOSX_LA_INTERFACE_${upper_LAI} "${upper_LAI} LA interface is selected" ON)
 
+###############################
+# NvToolExt
+###############################
+if (ENABLE_CUDA_NVTOOLSEXT)
```
Comment (TotoGaz) [inline src/cmake/thirdparty/SetupGeosxThirdParty.cmake:650]: Can you also check that `ENABLE_CUDA` is `ON?

File: src/coreComponents/common/TimingMacros.hpp [hunk 2]
```diff
@@ -45,40 +39,107 @@ namespace timingHelpers
...
+    YELLOW = 0xffffff00,
+    PURPLE = 0xffff00ff,
+    AQUA = 0xff00ffff,
+    RED = 0xffff0000,
+    WHITE = 0xffffffff
+  };
+
+  class NVTXScopeTracer {
+  public:
+    NVTXScopeTracer(const char* name, NVTXColors colorID) {
```
Comment (TotoGaz) [inline src/coreComponents/common/TimingMacros.hpp:61]:
```suggestion
    NVTXScopeTracer( const char* name, NVTXColors color ) {
```

File: src/coreComponents/common/TimingMacros.hpp [hunk 3]
```diff
@@ -45,40 +39,107 @@ namespace timingHelpers
...
+      nvtxRangePushEx(&eventAttrib);
+    }
+    ~NVTXScopeTracer() {
+      nvtxRangePop();
+    }
+  };
+
+}
+/// Mark a scope with NVTX with name and color (only a local helper should not be used elsewhere)
+#  define GEOSX_NVTX_MARK_SCOPE_2(name, color) timingHelpers::NVTXScopeTracer __FILE__ ## _ ## __LINE__ ## _ ## scopeTracer = timingHelpers::NVTXScopeTracer(name, color)
```
Comment (TotoGaz) [inline src/coreComponents/common/TimingMacros.hpp:78]:
```suggestion
#  define GEOSX_NVTX_MARK_SCOPE_COLORED(name, color) timingHelpers::NVTXScopeTracer __FILE__ ## _ ## __LINE__ ## _ ## scopeTracer = timingHelpers::NVTXScopeTracer(name, color)```

**Review by TotoGaz — APPROVED** [review]
Can you also add the appropriate doc about `ENABLE_CUDA_NVTOOLSEXT` in https://geosx-geosx.readthedocs-hosted.com/en/latest/docs/sphinx/buildGuide/BuildProcess.html ?
Can you also look at https://geosx-geosx.readthedocs-hosted.com/en/latest/docs/sphinx/developerGuide/Contributing/Caliper.html to see if there is something you can do at little cost?

**Comment by TotoGaz** [comment]
@rrsettgast despite minor doc problems, I think this is ready to go.
Is there anything required on the LLNL side?

**Comment by TotoGaz** [comment]
@XL64 are you sure it's ready to merge? There are still some questions here and there.
Plus some minor doc considerations.

File: src/coreComponents/common/TimingMacros.hpp (hunk 3 above)
Comment (XL64) [reply src/coreComponents/common/TimingMacros.hpp:78]: There are triple backquotes unwanted at the end + have to change the call also. I'll do it manually.

File: src/cmake/thirdparty/SetupGeosxThirdParty.cmake (hunk 1 above)
Comment (XL64) [reply src/cmake/thirdparty/SetupGeosxThirdParty.cmake:687]: Ok i'll do so.

**Comment by XL64** [comment]
Now it's ready

**Comment by TotoGaz** [comment]
@XL64 can you manage the conflict with `develop`, you're next to be merged.

**Pull request merged**